### ISSUE TITLE
Allow OVA/OVF deployment to be specified

### ISSUE THREAD

**Issue by adarobin** [hashicorp/terraform-provider-vsphere#1209]
### Description

Some OVF templates have different deployment sizes available that adjust the CPU, Memory, and Disk added to the machine at creation time.  Expose the ability to select from one of these deployment options.

I have working code and I'll submit a PR after I clean up the code a bit and make sure existing tests pass. Not sure exactly how to go about adding a new test for this though.

### Potential Terraform Configuration

```hcl
resource "vsphere_virtual_machine" "vcsa" {
  name             = "vcsa"
  datacenter_id    = data.vsphere_datacenter.dc.id
  folder           = var.folder_name
  resource_pool_id = data.vsphere_resource_pool.rp.id
  host_system_id   = data.vsphere_host.host.id
  datastore_id     = data.vsphere_datastore.datastore.id

  ovf_deploy {
    local_ovf_path    = "./vcenter.ova"
    disk_provisioning = "thin"
    ip_protocol       = "IPV4"
    deployment        = "small"

    ovf_network_map = {
      "Network 1" = data.vsphere_network.network.id
    }

    vapp {
      properties = {
          # a bunch of properties here
      }
    }
  }
}
```

### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

**Comment by ron4all** [comment]
this enhancement is shown as committed when look into the duplicate [https://github.com/hashicorp/terraform-provider-vsphere/pull/1215](https://github.com/hashicorp/terraform-provider-vsphere/pull/1215):

### Screenshot from 1215
![grafik](https://user-images.githubusercontent.com/26636908/102713169-6cb9b600-42c6-11eb-8942-977dcc3d6d7a.png)


### Screenshot from Changelog
* but nothing to find in the Changelog:
![grafik](https://user-images.githubusercontent.com/26636908/102713119-1c425880-42c6-11eb-97db-2cdc17abfdd0.png)

### Question
So it has been changed months ago but the change has been rolled-back later on?

**Comment by ron4all** [comment]
maybe it might make sense to change the thread-title
* from "Allow OVA/OVF `deployment` to be specified"
* to: "Allow OVA/OVF `DeploymentOption` to be specified"
?

As we see in having duplicates in this already - the term used in the OFV-Spec is "DeploymentOption"
* [Open Virtualization Format Specification, pdf](http://dmtf.org/sites/default/files/standards/documents/DSP0243_2.0.0.pdf) 
* [real-life example for Illustration](https://allones.de/2020/12/18/ovf-ova-properties/)

**Comment by ron4all** [comment]
> 
> 
> Don't you think that this would make is more confusing to use? We have a CPU and a memory property. If a user sets that, the expected outcome would be it's used. Of course behind the scenes the API might be different. But I don't see why a additional property DeploymentOptionSelection should be required just for this use case. The workflow would be "well, you can set the CPU using the CPU property, BUT if you use this, then you have to set it in a different way". From a user perspective this should all be abstrected away.

well, good question: I never tried to specify memory/cpu/operating-system when deploying OVA/OVFs since those parameters are specified by the DeploymentOption-Setting within the OVF-XML.
* so i expect, other people will find it intuitive, too, to not specify things manually which are already specified within the OVA/OVF-Appliance.

For my opinion, it would be more confusing, to have to manually add these (redundant) information.

On the other hand - I won't work anyway - so
* we technically have to pass the deployment-option to the VM
Why?
* manually specified parameters are overwritten by the "DeploymentOption"-Parameters specified within the OVA/OVF

### Example
I've been adding `num_cpus` and `memory` to my ".tf"-file:
```hcl
resource "vsphere_virtual_machine" "csr1kv-10" {
  for_each = local.sites
...
  num_cpus = 4
  memory   = 8192

ovf_deploy {
    # /vmfs/
    #local_ovf_path = "/volumes/DS-SOFTWARE/iso/cisco/csr1000v-universalk9.16.09.01.ova"
    local_ovf_path = "C:\\TEMP\\csr1000v-universalk9.16.09.01.ova"
    disk_provisioning = "thin"
    ovf_network_map = {
        "GigabitEthernet1" = data.vsphere_network.mgmt_lan.id,
,...
    }
  }
...
```

### Result
This is getting deployed:
![grafik](https://user-images.githubusercontent.com/26636908/102767371-bdd9b080-437f-11eb-88d0-82f9cfaf4ae4.png)
According to the "Default-DeploymentOption":
```Powershell
Key                : DeploymentOption
Value              :
DefaultValue       : 1CPU-4GB
OvfTypeDescription : string["1CPU-4GB", "2CPU-4GB", "4CPU-4GB", "4CPU-8GB"]
Description        : Small
                     Minimal hardware profile - 1 vCPU, 4 GB RAM
                     Medium
                     Medium hardware profile - 2 vCPUs, 4 GB RAM
                     Large
                     Large hardware profile - 4 vCPUs, 4 GB RAM
                     Large + DRAM Upgrade
                     Large hardware profile (requires purchase of DRAM upgrade SKU) - 4 vCPUs, 8 GB RAM
```
### Finding
So the DeploymentOption takes precedence.
### Conclusion
When not passing the DeploymentOption to the vCenter it uses the "default DeplyomentOption" - in my case the 1CPU-4G/Memory"-Setting.

**The DeploymentOption-Parameter needs to be passed from the .tf-Script to the VM/vCenter.**

**Comment by omniproc** [comment]
I look at a OVA/OVF spec as a "default spec" that comes with optional settings that the user can specify. In reality most OVA/OVF are simple dumps of an VM without any "option" to change cpu / memory. So while it IS possible to make cpu / memory and other settings customizable in an OVA/OVF in reality it's more common to simply set it at deploy time in the interface your hypervisor offers for that task. In the vCenter UI the "change / customize" hardware is a seperate tab for exactly that reason and it will override whatever was specified in the OVA/OVF.

**Comment by techBeck03** [comment]
Completely agree @ron4all that you wouldn't typically adjust the hardware specs outlined in the deployment option if a deployment option selection is defined.  The challenge here is you are creating the same type of resource in either case, a virtual machine.

## Proposed Solution

**IF** a deploymentOptionSelection is present in the OVF **THEN** the following should happen:

1) The state for this virtual machine should include an entry specifying it was deployed using a deployment option
2) All `ResourceType`(s) specified within the selected deployment option (or default if non is specified) should also be tracked with the vm state
3) By default, if a virtual machine is deployed using an OVF + deployment option then you are not allowed to change any resource definitions that overlap with `ResourceType`(s) specified in the OVF (which are now tracked within the resource state).  Attempting to do so should throw  a clear error  message explaining why it isn't allowed.
4) Optional:  Allow overrides to these `ResourceType`(s) with a specific flag.

**Comment by adarobin** [comment]
@ron4all in your example if you run apply again, you would see that it wants to adjust the cpu and memory of the machine to match what you have set.

With the code in #1215 you could specify the deployment option you would prefer, but if cpu and memory weren’t set to match the option you specified, a second apply would want to adjust it again.

I don’t see a way around not having to set cpu and memory along with the deployment option without removing the ability to make adjustments to the cpu and memory allocated to the machine which seems worse than the inconvenience of setting it twice.

**Comment by techBeck03** [comment]
@adarobin would my proposed solution not address what you are describing?

Thanks!

**Comment by ron4all** [comment]
> 
> 
> I look at a OVA/OVF spec as a "default spec" that comes with optional settings that the user can specify. In reality most OVA/OVF are simple dumps of an VM without any "option" to change cpu / memory. So while it IS possible to make cpu / memory and other settings customizable in an OVA/OVF in reality it's more common to simply set it at deploy time in the interface your hypervisor offers for that task. In the vCenter UI the "change / customize" hardware is a seperate tab for exactly that reason and it will override whatever was specified in the OVA/OVF.

I disagree.

We should distinguish between
a) there might be "exported VMs" which would contain no DeploymentOption - i think you're talking about those type of OVAs
b) there are "Professional Appliances" provided in OVA-Format by commercial IT-Vendors
... i could name more than a dozen virtual Appliances provided as OVA by VMware, Cisco, etc.

For Type (a)-OVAs it might be relevant to specify the VM-Settings manually, okay.

I'm talking about Type (b)-OVAs.
* here it might be important, to have the EXACT VM-Setup as specified in one of many DeploymentOption-Choices
* for example, to create a setup, which is valid to earn (paid) support from the OVA-Vendor.

If the vendor - i'm talking about professional OVA-Appliances - specifies several DeploymentOptions to choose from, no one will think about specifying those parameters manually:
* this is obsolete in this case.

My opinion:
* the DeploymentOption-Parameter might be optional, since Type (a)-OVAs don't require them
* **the DeplymentOption-Parameter should be possible to be specified**, since Type (b)-OVAs micht require them.

**Comment by ron4all** [comment]
> 
> 
> @ron4all in your example if you run apply again, you would see that it wants to adjust the cpu and memory of the machine to match what you have set.
> 
> With the code in #1215 you could specify the deployment option you would prefer, but if cpu and memory weren’t set to match the option you specified, a second apply would want to adjust it again.
> 
> I don’t see a way around not having to set cpu and memory along with the deployment option without removing the ability to make adjustments to the cpu and memory allocated to the machine which seems worse than the inconvenience of setting it twice.

It was just an example to show, that the DeploymentOption-Settings in the OVA override manual settings:
* this is absolutely ok
* people using these kiind of OVA won't override these VM-Settings manually, i assume

So i don't want to apply-twice to get my manually chosen settiings:
* i don't want to specify those settings manually
* this would be redundant spec - since the OVA contains those settings already

I need to specify the OVA DeploymentOption only to get a valid and supported VM.

I expect the code in [https://github.com/hashicorp/terraform-provider-vsphere/pull/1215](https://github.com/hashicorp/terraform-provider-vsphere/pull/1215) too solve the issue, absolutely.

**Comment by adarobin** [comment]
@ron4all in the case of the VCSA it is supported to adjust settings from the deployment option. For example, small might work based on the number of hosts you have but you might want to add additional cpu and ram to accommodate more concurrent users. VMware doesn’t require a VCSA to be exactly as configured by a deployment size.

@techBeck03 that could work, but it seems really complicated. Perhaps @bill-rich could weigh in on this?

**Comment by ron4all** [comment]
@adarobin 
> 
> 
> @ron4all in the case of the VCSA it is supported to adjust settings from the deployment option. For example, small might work based on the number of hosts you have but you might want to add additional cpu and ram to accommodate more concurrent users. VMware doesn’t require a VCSA to be exactly as configured by a deployment size.
> 

I'm no native speaker, maybe i'm misunderstanding - the deployment-process should contain two steps?
1. with "default" DeploymentOption-Parameters since other DeploymentOptions can't be selected
2. shutting down the VM, adjusting VM-parameters, bootup.

This would be possible.

Haven't you been the person inventing the solution in #1215? Would you mind, if i ask, what's wrong with the code in #1215?

**Comment by adarobin** [comment]
@ron4all I did write the code in #1215.  I was under the impression that you wanted to ignore the values set in cpu and memory if you set a deployment option for the OVF.  The way it is presently in #1215 is that you can deploy the machine with deployment option and the value of cpu and memory is ignored when the machine is initially created.  After that the values for CPU and memory must be set to match the sizes specified in the deployment option otherwise the machine will be changed.

**Comment by ron4all** [comment]
@adarobin
> 
> 
> @ron4all I did write the code in #1215. I was under the impression that you wanted to ignore the values set in cpu and memory if you set a deployment option for the OVF. The way it is presently in #1215 is that you can deploy the machine with deployment option and the value of cpu and memory is ignored when the machine is initially created. After that the values for CPU and memory must be set to match the sizes specified in the deployment option otherwise the machine will be changed.

I understand, and I'm sure, you've been spending much more time thinking about this than me.

To clarify what I want: I didn't even think about specifing cpu and memory since those parameters are set by the OVA/DeploymentOption.

What I really most like in Terraform is the idempotent behaviour - this feature is, what makes me want to migrate from my existing PowerShell-Scripts to Terraform, it will take time but that time would be worth spent, since the whole infrastructure-deployment-process would be much more relieable.

When applying the same .tf-config to an unchanged setup twice - and that would modify the setup - this would (imho) currupt this outstanding idempotency-feature.

Maybe that's about what @techBeck03 ment ("Proposed Solution") - i hope I understand correctly:
* the VM get's deployed as OVA - during deployment the DeploymentOption (if it exists) rules
* but when the VM is already deployed, all OVA-Stuff in the .tf-setup might need to be ignored since its only relevant to the first deployment and only the "normal" VM-parameters should be taken into consideration and checked, if changed
That would in my eyes enable an idempotent behaviour even for VMs which got deplyed by an OVA.
... otherwise would any change in the OVA-"app"-Parameters require a full redeployment of the VM to be idempotent, wouldn't it!?

**Comment by djpbessems** [comment]
I would like to point out that not every DeploymentOption will even have effect on cpu/ram sizing; the DMTF OVF specs (https://bit.ly/3oKNkws) make it clear that you can also use DeploymentOptions purely to select a subset of vApp properties.

See an example .ovf excerpt here: https://code.spamasaurus.com/djpbessems/Packer.Images/src/branch/Server2019/README.md

I just posted a question about this on HashiCorp's forum before I found this issue:
https://discuss.hashicorp.com/t/deploying-ovf-with-deploymentoption-ovf-configuration-vsphere-provider/20570

**Comment by tenthirtyam** [comment]
@adarobin, would you mind confirming that the scope of your enhancement request was  addressed in #1339?

Ryan

**Comment by tenthirtyam** [comment]
Marking this issue as closed based on the this being addressed in https://github.com/hashicorp/terraform-provider-vsphere/pull/1339.

If the issue persists with the solution provided and [the latest version of the provider](https://github.com/hashicorp/terraform-provider-vsphere/releases/latest), please open a new issue and reference this issue for additional context.

**Ryan Johnson**
Staff II Solutions Architect, VMware, Inc.
